src/cpp/src/text2image/schedulers/ddim.cpp

```diff
@@ -0,0 +1,184 @@
+    read_json_param(data, "steps_offset", steps_offset);
+    read_json_param(data, "prediction_type", prediction_type);
+    read_json_param(data, "thresholding", thresholding);
+    read_json_param(data, "dynamic_thresholding_ratio", dynamic_thresholding_ratio);
+    read_json_param(data, "clip_sample_range", clip_sample_range);
+    read_json_param(data, "sample_max_value", sample_max_value);
+    read_json_param(data, "timestep_spacing", timestep_spacing);
+    read_json_param(data, "rescale_betas_zero_snr", rescale_betas_zero_snr);
+}
+
+DDIMScheduler::DDIMScheduler(const std::string scheduler_config_path) 
+    : DDIMScheduler(Config(scheduler_config_path)) {
+}
+
+DDIMScheduler::DDIMScheduler(const Config& scheduler_config)
+    : m_config(scheduler_config) {
+
+    std::vector<float> alphas, betas;
+
+    using utils::linspace;
+
+    if (!m_config.trained_betas.empty()) {
+        betas = m_config.trained_betas;
+    } else if (m_config.beta_schedule == BetaSchedule::LINEAR) {
+        betas = linspace<float>(m_config.beta_start, m_config.beta_end, m_config.num_train_timesteps);
+    } else if (m_config.beta_schedule == BetaSchedule::SCALED_LINEAR) {
+        float start = std::sqrt(m_config.beta_start);
+        float end = std::sqrt(m_config.beta_end);
+        betas = linspace<float>(start, end, m_config.num_train_timesteps);
+        std::for_each(betas.begin(), betas.end(), [] (float & x) { x *= x; });
+    } else {
+        OPENVINO_THROW("'beta_schedule' must be one of 'LINEAR' or 'SCALED_LINEAR'. Please, add support of other types");
+    }
+
+    // TODO: Rescale for zero SNR
+    // if (m_config.rescale_betas_zero_snr) {betas = rescale_zero_terminal_snr(betas)}
+
+    std::transform(betas.begin(), betas.end(), std::back_inserter(alphas), [] (float b) { return 1.0f - b; });
+
+    for (size_t i = 1; i <= alphas.size(); i++) {
+        float alpha_cumprod =
+            std::accumulate(std::begin(alphas), std::begin(alphas) + i, 1.0, std::multiplies<float>{});
+        m_alphas_cumprod.push_back(alpha_cumprod);
+    }
+
+    m_final_alpha_cumprod = m_config.set_alpha_to_one ? 1 : m_alphas_cumprod[0];
+}
+
+void DDIMScheduler::set_timesteps(size_t num_inference_steps) {
+    m_timesteps.clear();
+
+    if (num_inference_steps > m_config.num_train_timesteps) {
+        OPENVINO_THROW("`num_inference_steps cannot be larger than m_config.num_train_timesteps");
+
+    }
+
+    m_num_inference_steps = num_inference_steps;
+
+    // TODO: add linspace and trailing
+    if (m_config.timestep_spacing == TimestepSpacing::LEADING) {
+        size_t step_ratio = m_config.num_train_timesteps  / m_num_inference_steps;
+
+        for (size_t i = num_inference_steps - 1; i != -1; --i) {
+            m_timesteps.push_back(i * step_ratio + m_config.steps_offset);
+        }
+
+    } else {
+        OPENVINO_THROW("'timestep_spacing' must be 'LEADING'. Please, add support of other types");
```

Comment:
I think it is better to be more specific and provide info which spacing specified in config in error message

Reply:
Supported all 3 timestep spacing types, other values will be thrown out at https://github.com/openvinotoolkit/openvino.genai/blob/e9d4caee4338de1cca678fb9406f011fd1f01e9c/src/cpp/src/text2image/schedulers/types.cpp#L60